### PR TITLE
Conflict check

### PR DESCRIPTION
**Pull request by mareq**

### CODE REVIEW

File: event-sauce/src/event_builder/conflict_event.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    /// Workaround method to get an entity ID out of entities when implementing
+    /// `ConflictEntityBuilder`
+    pub(crate) fn build_with_entity_id(self, entity_id: Uuid) -> Event<ConflictData<EDA, EDC>> {
+        Event {
+            id: Uuid::new_v4(),
+            event_type: String::from(self.payload.event_type()),
+            entity_type: EDA::Entity::entity_type(),
+            entity_id,
```
Comment (jamwaffles) [inline event-sauce/src/event_builder/conflict_event.rs:41]: I think the entity ID can be reused from `self.payload`, no? Then the `entity_id` argument to this method can be removed.

```suggestion
            self.payload.applied_event.entity_id,
```

File: event-sauce/src/lib.rs [hunk 2]
```diff
@@ -69,6 +69,61 @@ pub trait EventData: Serialize + Sized {
...
+where
+    EDA: EventData,
+{
+    /// Check if applying `self` is in conflict with an already `applied_event`.
+    ///
+    /// Returns either `Ok(EventData)` in case of no conflicts or [`ConflictData`] with `Self` in
+    /// place of `EDC`, describing the conflict otherwise.
+    ///
+    /// This function will be called during [`UpdateEntityBuilder::try_update`](trait.UpdateEntityBuilder.html#method.try_update).
+    fn conflict_check(self, applied_event: &Event<EDA>) -> Result<Self, ConflictData<EDA, Self>>;
```
Comment (jamwaffles) [inline event-sauce/src/lib.rs:83]: This method is performing the act of checking for a conflict, so makes more sense to call it `check_conflict`:

```suggestion
    fn check_conflict(self, applied_event: &Event<EDA>) -> Result<Self, ConflictData<EDA, Self>>;
```

File: event-sauce/src/lib.rs [hunk 3]
```diff
@@ -235,6 +309,28 @@ where
...
 }
 
+/// A wrapper trait around [`AggregateConflict`] to handle event-sauce integration boilerplate
+pub trait ConflictEntityBuilder<EDA, EDC>: AggregateConflict<EDA, EDC> + Entity
+where
+    EDA: EventData,
+    EDC: EventData,
+{
+    /// Conflict the entity with an event
+    fn try_conflict<B>(
```
Comment (jamwaffles) [inline event-sauce/src/lib.rs:319]: Wouldn't this make more sense as `try_flag_conflict` or `try_store_conflict` or something? We're not attempting to _cause_ a conflict, but store its result here, right?

File: event-sauce/tests/action_builder.rs [hunk 4]
```diff
@@ -25,32 +25,46 @@ impl AggregateAction<UserEventData> for User {
...
-                        // TODO: Better error variant
-                        .map_err(|_e| EventError::Infallible())?;
+                    let create_event =
+                        event
+                            .clone()
+                            .try_into_variant::<UserCreated>()
+                            .map_err(|_| {
+                                EventError::ConversionError(
+                                    "Event<UserEventData>",
+                                    "Event<UserCreated>",
```
Comment (jamwaffles) [inline event-sauce/tests/action_builder.rs:35]: Not something to change for this PR, but for future code you could use [`type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) in these cases.

**Review by jamwaffles — CHANGES_REQUESTED** [review]
Some small naming nits, otherwise LGTM

File: event-sauce/tests/action_builder.rs (hunk 4 above)
Comment (mareq) [reply event-sauce/tests/action_builder.rs:35]: Ah, I was looking for something like this - thank!

File: event-sauce/src/lib.rs (hunk 3 above)
Comment (mareq) [reply event-sauce/src/lib.rs]: I see what you had in mind, when speaking about having to think about the naming - both of those two suggestions are bit weird names, but it is indeed challenging to come up with better ones (I do not have any better ideas). So `try_flag_conflict` it is.

**Comment by mareq** [comment]
Ready for merge tomorrow morning :crossed_fingers:

**Review by Istar-Eldritch — APPROVED** [review]

**Review by jamwaffles — APPROVED** [review]

**Pull request merged**